v2/api/network/customizations/private_endpoints_extensions.go

```diff
@@ -60,3 +67,99 @@ func (extension *PrivateEndpointExtension) PostReconcileCheck(
 
 	return extensions.PostReconcileCheckResultSuccess(), nil
 }
+
+func (extension *PrivateEndpointExtension) ExportKubernetesResources(
+	ctx context.Context,
+	obj genruntime.MetaObject,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger,
+) ([]client.Object, error) {
+	// This has to be the current hub storage version. It will need to be updated
+	// if the hub storage version changes.
+	endpoint, ok := obj.(*network.PrivateEndpoint)
+	if !ok {
+		return nil, errors.Errorf("cannot run on unknown resource type %T, expected *network.PrivateEndpoint", obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not
+	var _ conversion.Hub = endpoint
+
+	hasIpConfiguration := configMapSpecified(endpoint)
+	if !hasIpConfiguration {
+		log.V(Debug).Info("no configmap retrieval to perform as operatorSpec is empty")
+		return nil, nil
+	}
+
+	if endpoint.Status.NetworkInterfaces == nil || len(endpoint.Status.NetworkInterfaces) == 0 {
+		log.V(Debug).Info("no configmap retrieval to perform as there is no NetworkInterfaces attached")
+		return nil, nil
+	}
+
+	nicId, err := arm.ParseResourceID(*endpoint.Status.NetworkInterfaces[0].Id)
+	if err != nil {
+		return nil, err
+	}
+
+	// The default primary ip configuration for PrivateEndpoint is on NetworkInterfaceController. Hence, we fetch it from there.
+	var interfacesClient *armnetwork.InterfacesClient
+	interfacesClient, err = armnetwork.NewInterfacesClient(nicId.SubscriptionID, armClient.Creds(), armClient.ClientOptions())
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create new PrivateEndpointsClient")
+	}
+
+	var resp armnetwork.InterfacesClientGetResponse
+	resp, err = interfacesClient.Get(ctx, nicId.ResourceGroupName, nicId.Name, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed getting PrivateEndpoint")
+	}
+
+	configsByName := configByName(resp)
+	configs, err := configMapToWrite(endpoint, configsByName)
+	if err != nil {
+		return nil, err
+	}
+
+	return configmaps.SliceToClientObjectSlice(configs), nil
+}
+
+func configByName(resp armnetwork.InterfacesClientGetResponse) map[string]string {
```

Comment:
minor: Could you combine the `configByName` and `configMapToWrite` sections here?

I think in some secrets handlers we do this pattern because the actual data comes from a `map[string]string` to start, but here it's well-typed so it feels like just passing `armnetwork.InterfacesClientGetResponse` to `configMapToWrite` is OK?

Though actually, you shouldn't pass the `GetResponse`, if you look inside it there should be a resource type embedded and that's what you wanna pass. You don't need the HTTP junk that's with it on the `GetResponse`.

Reply:
I followed this approach to make the code easily extendable in the future if we plan on adding more config exports. That way it's cleaner to just have all the fetching logic in one method and writing logic in another.

Reply:
The thing I think could be improved about the way it is now is that you fetch from the typed structure into a `map[string]string`, with a somewhat magical key of `primaryNICPrivateIPAddress`, and then you write by reading that somewhat magical key again.

My argument is that it's no clearer to read it from a `map[string]string` than it is to read it from a `network.PrivateEndpoint`, and if in the future we need to read other things from the same `network.PrivateEndpoint` there's no advantage. Only if we were reading from multiple places might there be a slight advantage, but even then it's only slight because while you avoid needing N parameters for the `configMapToWrite` function, you still need N parameters for the `configByName` function so it's not really all that much cleaner/clearer?

It feels like the `map[string]string` is adding a step that doesn't really have a ton of value now. I'd argue rather than position us for a possible future where things are more complicated we just keep it simple now and could split into 2 functions in the future if/when the need arises. YAGNI and all that, I guess.

Alternatively, if you want the structure of the split maybe you could accomplish it by still having 1 function, but it constructs the `collector` and passes it to a function that extracts the values and writes it - that way there's no intermediate map it's just straight from the struct to the collector? Then if there were other sets to collect you'd have another function for those and pass the same collector down to it too?